### ISSUE TITLE
Add `preferred_element_type` to `nnx.Linear` and other relevant layers

### ISSUE THREAD

**Issue by manuelgloeckler** [google/flax#4890]
### 🚀 Feature request

A common mixed-precision pattern is to perform *heavy* computations (e.g., matrix multiplications in `Linear` layers) in `jnp.bfloat16` or `jnp.float16`, but then accumulate the end result in `jnp.float32` for subsequent operations, such as activation functions or normalization layers, to preserve precision.

Right now, achieving this behavior in Flax requires overriding `dot_general` or `conv_general_dilated` to set the `preferred_element_type`, which is cumbersome. Most of the relevant keyword arguments (e.g., `dtype`, `precision`) are already exposed in `__init__`, so the only missing piece is **`preferred_element_type`**.

---

### 💡 Proposed solution

Expose a `preferred_element_type` argument in layers like `nnx.Linear` and pass it through to the underlying JAX calls (`lax.dot_general`, `lax.conv_general_dilated`, etc.).  This would make mixed-precision setups significantly easier and avoid custom overrides.

---

### 📋 Minimal example

The example below demonstrates why `preferred_element_type` is necessary for performing computations in the intented precision with the expected runtime improvements. There might be a different way to achive the same, but I do not see it.

```python
import jax
import jax.numpy as jnp
from flax import nnx
from functools import partial

linear = jax.jit(nnx.Linear(5096, 5096, rngs=nnx.Rngs(0), precision="BF16_BF16_F32", dtype=jnp.bfloat16))

linear2 = jax.jit(
    nnx.Linear(
        5096, 5096, rngs=nnx.Rngs(0), precision="BF16_BF16_F32", dtype=jnp.bfloat16,
        dot_general=partial(jax.lax.dot_general, preferred_element_type=jnp.float32)
    )
)

linear3 = jax.jit(lambda x: linear(x).astype(jnp.float32))  # cast afterwards (may be optimized away by XLA)
linear4 = jax.jit(nnx.Linear(5096, 5096, rngs=nnx.Rngs(0)))  # default F32

x = jnp.ones((1024, 5096), dtype=jnp.float32)

# Precision
out = linear(x)    # jnp.bfloat16 (forced downcast)
out2 = linear2(x)  # jnp.float32
out3 = linear3(x)  # jnp.float32
out4 = linear4(x)  # jnp.float32

print(jnp.abs(out4 - out).mean())   # 0.00178 (worse precision)
print(jnp.abs(out4 - out2).mean())  # 0.00131 (better precision)
print(jnp.abs(out - out3).mean())   # 0.0

# Runtime (similar for BF16, much faster than F32)
# %timeit -> 369 μs ± 2.98 μs   (linear)
# %timeit -> 370 μs ± 0.64 μs   (linear2, with preferred_element_type)
# %timeit -> 388 μs ± 2.79 μs   (linear3)
# %timeit -> 683 μs ± 3.11 μs   (linear4, default F32)

**Comment by vfdev-5** [comment]
Thanks for the feature request @manuelgloeckler ! 
It is true that some kwargs (`precision`) of `lax.dot_linear` are exposed in `nnx.Linear` constructor:
```python
    y = self.dot_general(
      inputs,
      kernel,
      (((inputs.ndim - 1,), (0,)), ((), ())),
      precision=self.precision,
    )
```

Exposing `preferred_element_type` and passing it into the underlying function would require updates in:
- nnx.Linear
- nnx.LinearGeneral
- nnx.Einsum
- nnx.Conv
- nnx.ConvTranspose

I'm not against adding it. @cgarciae @IvyZX what do you think?

**Comment by cgarciae** [comment]
@manuelgloeckler @vfdev-5 I think exposing `preferred_element_type` makes sense and exposing it in the APIs mentioned by @vfdev-5 if either or you want to take this issue.
